external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -225,6 +237,23 @@ private long doSeek(TopicPartition tp, OffsetAndMetadata committedOffset) {
         }
     }
 
+    /**
+     * Checks If {@link OffsetAndMetadata} was committed by this topology, either by this or another spout instance.
+     * This info is used to decide if {@link FirstPollOffsetStrategy} should be applied
+     *
+     * @param committedOffset {@link OffsetAndMetadata} info committed to Kafka
+     * @return true if this topology committed this {@link OffsetAndMetadata}, false otherwise
+     */
+    private boolean isOffsetCommittedByThisTopology(OffsetAndMetadata committedOffset) {
+        try {
+            return committedOffset != null && JSON_MAPPER.readValue(committedOffset.metadata(), KafkaSpoutMessageId.class)
```

Comment:
Nit: Move the null check outside this method, it's already checked before using this function at one call site, and in the other I don't think it makes the code more readable that it's moved in here. Also consider putting the message id in a variable before doing the .getTopologyId, that way it's easier to tell what's happening in case we get an NPE from this method.

Reply:
Also I think this will crash the spout when starting up on a topic that already has commits from before this change. We should probably warn in the log and pick some reasonable default instead of erroring out if there's no message id in the metadata.

Reply:
Agree. Up to this point this was a bug. I guess we should default to the previous behavior before this fix, both in 1.x and 2.x. What do you think?

Reply:
I'd probably just return false from this method. Since the topology id is new for each deployment we know for certain that if there's no message id in the meta, then it wasn't committed by this topology.

Reply:
That's what I did

Reply:
I'm a little unsure what you mean. Won't this currently throw an IOException (or NPE) when you start from a committed offset that doesn't have a KafkaSpoutMessageId in the metadata?

Reply:
It returns false if it throws exception when attempting to deserialize OffsetAndMetadata committed using a previous version of Storm.